services/src/main/java/io/grpc/services/HealthCheckingLoadBalancerProvider.java

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright 2018 The gRPC Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.services;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+
+import io.grpc.ExperimentalApi;
+import io.grpc.LoadBalancer;
+import io.grpc.LoadBalancer.Factory;
+import io.grpc.LoadBalancer.Helper;
+import io.grpc.LoadBalancerProvider;
+import io.grpc.internal.ExponentialBackoffPolicy;
+import io.grpc.internal.TimeProvider;
+
+/**
+ * A LoadBalancerProvider that wraps another provider and adds <a
+ * href="https://github.com/grpc/proposal/blob/master/A17-client-side-health-checking.md">client-side
+ * health checking</a>.
+ */
+@ExperimentalApi("https://github.com/grpc/grpc-java/issues/5025")
+public abstract class HealthCheckingLoadBalancerProvider extends LoadBalancerProvider {
```

Comment:
I don't think this should be a provider and I don't think it should be extended. Extending means you can't mix-and-match other similar solutions. Instead, I think each LB Provider that wants this should call a utility during `newLoadBalancer()`.

Reply:
I have change the PR to expose a utility instead. PTAL.